### DISCUSSION TITLE
Encoding Dolby Vision - Re-encoding (HDR10(+)/DV, HLG, Ngx-TrueHDR) for playback on Fire TV QLED and possibly other TVs

### DISCUSSION THREAD

**Discussion by noblemd00** [staxrip/staxrip#1179] [Show and tell]
Note: WORK IN PROGRESS - Check back often for updates - StaxRip updates may result in changes to settings.
<img width="553" alt="240609 110227" src="https://github.com/staxrip/staxrip/assets/16325865/50665b3a-138d-48fb-9ffb-e2f7e594eab8">

(Edited for my grammatical errors, poor typing, and updated information) 
My solution- experimentation - Additional information or corrections from readers welcome.

VUI SETTINGS MATRIX 
parameters to re-encode
HDR(10+)/Dolby Vision/HLG/Ngx-TrueHDR 
---
Encoder Options - VPP - VUI Settings

<img width="765" height="522" alt="_  Re-encode VUI Parameters" src="https://github.com/user-attachments/assets/e5ee9a70-305b-46c2-9c11-e238c0288867" />

Ngx-TrueHDR special options: 
---
Encoder Options - Basic - Output Depth must = 10-Bit
Encoder Options - VUI - Master Display = 
G(13250,34500)B(7500,3000)R(34000,16000)WP(15635,16450)L(10000000,1)
Encoder Options - VUI2 - Maximum CLL = 1000 and Maximum FALL = 300

2024.1002 - Moved VUI settings Matrix to the top.
2024.1031 - add Ngx-TrueHDR setting Encoder Options - Basic - Output Depth = 10-Bit
2024.10.03 - Added Ngx-TrueHDR information (NVIDIA AI to convert SDR to HDR - see Conversion Matrix picture below...)
Ngx-TrueHDR Requierments:
```
Windows PC with Windows 10 v1709 (Fall 2017 Creators Update 64-bit) or newer
NVIDIA RTX GPU (Quadro, Titan or GeForce)
Latest [NVIDIA Graphics Driver](https://www.nvidia.com/download/index.aspx), minimum 410.63.
```

2024.07.30 - KODI version 21+ now plays MKV Dolby Vision videos properly.

Updated - 20240313 - StaxRip (v2.36.2+?) Dolby Vision/HDR compatible settings. & HDR Multi Tool update v.2.1.0 (DV Cropping)

**### May work for other Dolby Vision TVs (not tested)**

Dolby Vision Profile Variants:
8.1: Fallback to HDR10.
8.2: Fallback to SDR (Standard Dynamic Range).
8.4: Fallback to HLG (Hybrid Log-Gamma)

1. I use Kodi installed on Fire TV to playback videos from my NAS.
2. KODI version 21 now properly plays MKV Dolby Vision Videos properly.
3. When re-encoding for archiving (storage on NAS), I use NVEncC (Nvidia) - H265  (speedy encoding, small file size, and video quality is excellent on 65 inch).

<img width="313" alt="Playing Dolby Vision" src="https://github.com/staxrip/staxrip/assets/16325865/e11d124f-7fda-447a-aa9c-dc00df8e1457">

***********************************************************************************************

The settings I use to get Dolby Vision playback on Fire TV QLED...
----

**Update: StaxRip v.2.33(+) should extract HDR10+ and Dolby Vision metadata automatically and insert the RPU/JSON files path/name in the input boxes.**

1.  Determine the quality/size you want for the re-encoded file - in StaxRip set the appropriate video/audio quality/size settings...
NOTE: StaxRip v.2.32 (+) requires VBR to be used . 

<img width="494" alt="_ Quality Settings" src="https://github.com/staxrip/staxrip/assets/16325865/737f5916-0f13-4dcf-b978-d665977cbf4f">

2. Drag-n-drop the file you want to re-encode into the main StaxRip Window and remember to change VUI TRANSFER back to "UNDEFINED" - for a Dolby Vision only and DV/HDR compatible file.

**********************************************************************************************
Update:20241003
Updated matrix info to include Ngx-TrueHDR settings (convert SDR to HDR using NVIDIA AI - see notes below)

Update 20240501
KODI version 21+ now properly plays Dolby Vision MKVs, I will update settings below. 
StaxRip defaults may now be used to re-encode Dolby Vision videos using KODI.

Updated 20240801 
Note 1: Kodi v.21 now plays Dolby Vision MKV videos on Fire TV
Note 2: Videos may be re-encoded as MKV or MP4

SETTINGS for RE_ENCODING (HDR, DV-HDR compatible -use StaxRip defaults with changes as noted in diagram below).

To re-encode DoVi only (Profile 5), change Options-Video-RPU Conversion Mode (picture below) and VUI settings listed.
This setting should also be used to re-encode Dolby Vision/HDR10+ encoded videos.
<img width="528" alt="240730 172408" src="https://github.com/user-attachments/assets/bb490721-ccf3-4621-b1e4-7da86f9c8781">
And you may have to enter VUI Dolby Vision RPU PATH and FILE NAME after the source is loaded - check other VUI settings as well.

To re-encode HLG videos use settings below.
NOTE:  HLG settings can be used to re-encode SDR videos as HLG
            BT 2020-12 entries can also be BT 2020-10 and vise-versa

Update: 20241009 - Default settings for most videos - SEE CAUTION - some videos need lower Max Luminance (500-800)
Update: 20241003 - updated matrix info - Ngx-TrueHDR settings
Update: 20240811 - HDR10+ Matrix info change Transfer = SMPTE 2084 - some videos require ATC SEI = Auto 
Update: 20240813 - HDR10+ (only) - Try ATC SEI = UNKNOWN first this seems to work best for most files. 

Note: I am continuing to check different HDR10+videos to see which settings give the best color, proper brightness, and yields correct MediaInfo data. Always check re-encoded video MediaInfo output for proper encoding.

Notes: 
Dolby Vision Profile 5 (DV only), StaxRip 2.41.7 (and before): You must set the RPU File/Path manually in VUI after the source video is loaded.
Options-Video-RPU Conversion Mode - must be set before source load - or saved/set in a template.

Use Ngx-TrueHDR to convert SDR videos to HDR. 
-----
Encoder Options - Basic - Output Depth must = 10-Bit
Encoder Options - VPP - Ngx-TrueHDR - Select checkbox: AI enhanced SDR to HDR conversion using RTX Video SDK
After further experimentation ... (UPDATE)
Default settings work in most situations. (For more vibrant colors set Saturation = 125-135)
Encoder Options - VUI - Master Display = 
G(13250,34500)B(7500,3000)R(34000,16000)WP(15635,16450)L(10000000,1)
Encoder Options - VUI2 - Maximum CLL = 1000 and Maximum FALL = 300

**********************************************************************************************
If you have an Ngx-TrueHDR capable NVIDIA video card - see NGX info above.
Note: DV+HLG re-encoded from DV/HDR(10+) videos (Transfer characteristics HLG/PQ).
All yield proper MediaInfo HDR Format entries and proper HDR (HDR only TV) and Dolby Vision (DV-HDR TV) playback.

You can automatically change the Transfer entry (and other settings) by creating Event Commands.

I name my templates for re-encoding with various names depending on the source encoding,

Dolby Vision / HDR(10+) with "HDV" in the name (i.e. Movie.4k.HDV.2.mp4.srip)
Dolby Vision only (purple/green tint) with "DV" in the name (i.e. Movie.4k.DV.2.mp4.srip)
HDR(10+) only with "HDR" in the name (i.e. Movie.4k.HDR.srip)
HLG with "HLG" in the name (i.e. Movie.4k.HLG.srip)
SDR with "SDR" in the name (i.e. Movie.4k.SDR.srip)

<img width="473" alt="240729 161421" src="https://github.com/user-attachments/assets/ff308e20-794b-441b-b3a0-4862a0aeccf7">

Essentially, the event command checks to see...
1. If the Template Name Contains "HDV" THEN
2. Change the Transfer setting to "bt2020-12"
3. The parameters are case sensitive - must be entered with the proper case "bt2020-12" NOT "BT2020-12"
5.  To enter multiple Command Line items such as required for HLG list one after the other such as:
      --transfer  "bt2020-12" --colormatrix "bt2020nc" --colorprim "bt2020" --atc-sei "arib-std-b67"

See the HELP (right-click the VUI entry names) for the parameters to use in the Command Line entry

<img width="606" alt="240729 162401" src="https://github.com/user-attachments/assets/a1038c04-9c21-4347-9c9d-0058cbe575de">

**********************************************************************************************

**** Dolby Vision only 
----
Profile 5 (Green/Purple tint) video, 

NOTE: Kodi v.21+ will play Dolby Vision videos encoded as MKV or MP4

Use the VUI settings in the diagram above.

---------------------------------------

NOTE 1: After Source loaded (if altered) change TRANSFER back to UNDETERMINED.
NOTE 2: NVEncC - VUI Dolby Vision RPU PATH/FILENAME (if blank) may have to be selected manually (click ... ).
NOTE 3: v.2.36+ Options-Video-RPU Conversion Mode should be 0: Parses RPU, rewrites it untouched or Untouched.

Update: StaxRip v.2.36 - Tested several different settings (see picture below) playback in Dolby Vision works properly. v.2.36 (+) StaxRip either ignores the settings or properly uses the settings.

<img width="321" alt="240225 145811" src="https://github.com/staxrip/staxrip/assets/16325865/1d7dc9fc-68db-4647-95f3-2ff396461bc3">

NOTE 4: HDR10 Info File entry should be blank.
NOTE 5: VUI TRANSFER may need changing back to UNDEFINED after Source loaded.

*************************************

**** Dolby Vision/HDR compatible (Updated 20240313)
----
NOTE 1: VUI Transfer must be UNDEFINED for DoVi playback on Fire TV QLED
NOTE 2: KODI v. 21+ Dolby Vision Videos may be encoded as MKV or MP4.
NOTE 3: Instead of "copy" you can use the HDR10(+) info .json metafile if available (StaxRip v2.32+ should auto-extract .json file). 
NOTE 4: After Source loaded VUI TRANSFER (if reset to SMPTE 2084) must be changed to UNDEFINED.

UPDATE: The below settings are the only ones that yield proper brightness with both Dolby Vision playback and HDR(10+) playback  Settings other than below result in HDR playback too bright.

Some DV/HDR videos require this change:
<img width="320" alt="AVS Filters" src="https://github.com/staxrip/staxrip/assets/16325865/e5716213-8c89-4db6-bd9f-c2b77b56ff7e">

*************************************

***** HDR10(+) only
----
These settings give HDR10+ (if the original was HDR10+) - Extract HDR10 metedata.json (or use "copy") and re-encode using these settings (StaxRip v2.32+ should auto-extract .json file).  
NOTE 1: HDR10(+) only files can be MKV or MP4.
NOTE 2: Note the Colorrange "Limited" setting.

*************************************

All other StaxRip settings are default or as pictured below...
----
<img width="283" alt="240106 155910" src="https://github.com/staxrip/staxrip/assets/16325865/61527f5e-ad03-414b-bc26-120b85382551">

HDR-Multi-Tool-GUI can still be used to manually extract Dolby Vision RPU BIN and HDR JSON file metadata. Parse the Dolby Vision RPU (and HDR Metatdata) files. Save the RPU (and Metadata) files to the StaxRip Source file temp folder. 
https://github.com/jlw4049/HDR-Multi-Tool

NOTE 1: Some files require "CROPPED ENCODE" to be checked to get proper size video playback. Always check playback for proper cropping on the Media Server/video program (Kodi, Plex, etc.). 
NOTE 2: HDR Multi Toll update 2.1.0 - DV Cropping also requires cropping values to be entered. 
NOTE 3: CHECK MARK "CROPPED ENCODE" to ensure aberrant Dolby Vision cropping does not occur. 

<img width="516" alt="HDRMultiTool_Arrows" src="https://github.com/staxrip/staxrip/assets/16325865/8081a502-7074-46dc-b248-c1a250ef79ef">

v.2.1.0
<img width="514" alt="240314 100920" src="https://github.com/staxrip/staxrip/assets/16325865/413d7e3e-8ad9-4cdf-a048-24ee0720dbbe">

<img width="517" alt="HDRMetadata" src="https://github.com/staxrip/staxrip/assets/16325865/284f1821-c6da-419c-bb7b-d231ae317267">

**Comment by jessielw** [comment]
I'm the author of the UI for the hdr tools. Note there is a newer version written in a completely modern framework here. 

https://github.com/jlw4049/HDR-Multi-Tool

Additionally while some of the information is correct above you should be extracting the rpu as profile `8.1`. This is the only valid profile to be injected into for playback in mkv/mp4 containers.

**Comment by subnrg** [comment]
Thanks @jessielw this was super helpful and works well for me except on DV 7.6 profile.  I am normally taking a 4K DV remux down to 1080p nvenc 6MB VBR 10bit 4.1 AAC DV 8.1.  I can't seem to figure out the magic formula when the source is DV 7.6 however.  Any ideas?

**Reply by jessielw** [reply]
Should still just be **8.1** for the output. Use the tool and it'll work just fine.

**Comment by DaMacFunkin** [comment]
Hi, I have been trying NGX HDR conversion from 8bit SDR rec709 UHD source to 10 Bit HDR bt2020, I can't find a way to De-saturate the Red, even if I set the Saturation to 0. The best results I have are with setting Contrast 0, Saturation 0, Middle Grey 43, Max Lum 1000. (I had to set contrast to 0 it was to dark at any other setting).
Does anybody have a full list of settings to achieve good results please?

**Reply by Dendraspis** [reply]
Regarding the encoder settings this is what you are looking for:
https://github.com/rigaya/NVEnc/blob/master/NVEncC_Options.en.md#--vpp-ngx-truehdr-param1value1param2value2

Beside that you can tweak the values beforehand via filters in AviSynth/VapourSynth when NVEncC is not supposed to read the source file as well.

**Comment by alexeik** [comment]
Are there possibile to add HDR10 from ngx-truehdr as option to video?
now i can only make only HDR10 video file.
Need to switch between original bt709 and HDR10+ bt2020

**Reply by Dendraspis** [reply]
Everything related can be found in the `NVEncC` documentation:
https://github.com/rigaya/NVEnc/blob/master/NVEncC_Options.en.md#--vpp-ngx-truehdr-param1value1param2value2

Adding HDR10+ data should not work as it is always based on a given HDR10 data.

**Reply by alexeik** [reply]
Nice moment to notice.
I thougth ngx true hdr can add HDR10 layer to video.
